File: ina_1sem/labor/6_lab/1/queens.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
void generate_permutations(int arr[], int n, int *solutions) {
    while (true) {
        bool is_valid = 1;
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                if (abs(arr[i] - arr[j]) == abs(i - j)) {
                    is_valid = 0;
                    break;
                }
            }
            if (!is_valid) break;
        }

        if (is_valid) {
            for (int i = 0; i < n; ++i) {
                printf("%d ", arr[i] + 1);
            }
            printf("\n");
            (*solutions)++;
        }

        int i = n - 2;
        while (i >= 0 && arr[i] > arr[i + 1]) {
            --i;
        }
        if (i < 0) break;

        int j = n - 1;
        while (arr[j] <= arr[i]) {
            --j;
        }

        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;

        //swap(arr[i], arr[j]);

        // reverse array from (i + 1) to (n-1)
        int left = i + 1, right = n - 1;

        while (left < right) {
            temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            ++left;
            --right;
        }
    }
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <n>\n", argv[0]);
        return 1;
    }

    int n = atoi(argv[1]);
    if (n < 1) {
        printf("Please enter a positive integer for n.\n");
        return 1;
    }

    int *arr = malloc(n * sizeof(int));
    if (arr == NULL) {
        printf("Memory allocation failed.\n");
        return 1;
    }

    // Init array
    for (int i = 0; i < n; ++i) {
        arr[i] = i;
    }

    int solutions = 0;
    generate_permutations(arr, n, &solutions);

    printf("Number of solutions: %d\n", solutions);

    free(arr);
    return 0;
}

File: ina_1sem/labor/6_lab/3/queens.py
def is_valid_permutation(arr):
    n = len(arr)
    for i in range(n):
        for j in range(i + 1, n):
            if abs(arr[i] - arr[j]) == abs(i - j):
                return False
    return True

def generate_permutations(n):
    arr = list(range(n))
    solutions = 0

    while True:
        if is_valid_permutation(arr):
            print([x + 1 for x in arr])
            solutions += 1

        i = n - 2
        while i >= 0 and arr[i] > arr[i + 1]:
            i -= 1

        if i < 0:
            break

        j = n - 1
        while arr[j] <= arr[i]:
            j -= 1

        arr[i], arr[j] = arr[j], arr[i]
        arr[i + 1:] = reversed(arr[i + 1:])

    return solutions

def main():
    n = int(input("Enter the value of n: "))
    solutions = generate_permutations(n)
    print(f"Number of solutions: {solutions}")

if __name__ == "__main__":
    main()

